runtime/lua/vim/lsp/codelens.lua

```diff
@@ -0,0 +1,223 @@
+    api.nvim_buf_clear_namespace(bufnr, ns, i, i + 1)
+    local chunks = {}
+    for _, lens in pairs(line_lenses or {}) do
+      local text = lens.command and lens.command.title or 'Unresolved lens ...'
+      table.insert(chunks, {text, 'LspCodeLens' })
+    end
+    if #chunks > 0 then
+      api.nvim_buf_set_virtual_text(bufnr, ns, i, chunks, {})
+    end
+  end
+end
+
+
+--- Store lenses for a specific buffer and client
+---
+---@param lenses table of lenses to store (`CodeLens[] | null`)
+---@param bufnr number
+---@param client_id number
+function M.save(lenses, bufnr, client_id)
+  local lenses_by_client = lens_cache_by_buf[bufnr]
+  if not lenses_by_client then
+    lenses_by_client = {}
+    lens_cache_by_buf[bufnr] = lenses_by_client
+    local ns = namespaces[client_id]
+    api.nvim_buf_attach(bufnr, false, {
+      on_detach = function(b) lens_cache_by_buf[b] = nil end,
+      on_lines = function(_, b, _, first_lnum, last_lnum)
+        api.nvim_buf_clear_namespace(b, ns, first_lnum, last_lnum)
+      end
+    })
+  end
+  lenses_by_client[client_id] = lenses
+end
+
+
+--@private
+local function resolve_lenses(lenses, bufnr, client_id, callback)
+  lenses = lenses or {}
+  local num_lens = vim.tbl_count(lenses)
+  if num_lens == 0 then
+    callback()
+    return
+  end
+
+  --@private
+  local function countdown()
+    num_lens = num_lens - 1
+    if num_lens == 0 then
+      callback()
+    end
+  end
+  local ns = namespaces[client_id]
+  local client = vim.lsp.get_client_by_id(client_id)
+  for _, lens in pairs(lenses or {}) do
+    if lens.command then
+      countdown()
+      goto continue
+    end
+
+    client.request('codeLens/resolve', lens, function(_, _, result)
+      if result and result.command then
+        lens.command = result.command
+        -- Eager display to have some sort of incremental feedback
+        -- Once all lenses got resolved there will be a full redraw for all lenses
+        -- So that multiple lens per line are properly displayed
+        api.nvim_buf_set_virtual_text(
+          bufnr,
+          ns,
+          lens.range.start.line,
+          {{ lens.command.title, 'LspCodeLens' },},
+          {}
+        )
+      end
+      countdown()
+    end, bufnr)
+
+    ::continue::
+  end
+end
+
+
+--- |lsp-handler| for the method `textDocument/codeLens`
+---
+function M.on_codelens(err, _, result, client_id, bufnr)
+  assert(not err, vim.inspect(err))
+
+  M.save(result, bufnr, client_id)
+
+  -- Eager display for any resolved (and unresolved) lenses and refresh them
+  -- once resolved.
+  M.display(result, bufnr, client_id)
+  resolve_lenses(result, bufnr, client_id, function()
+    M.display(result, bufnr, client_id)
+  end)
+end
+
+
+--- Refresh the codelens for the current buffer
+---
+--- It is recommended to trigger this using an autocmd or via keymap.
+---
+--- <pre>
+---   autocmd BufEnter,CursorHold,InsertLeave <buffer> lua vim.lsp.codelens.refresh()
+--- </pre>
+---
```

Comment:
I'm not sure if it is good to call this like that without some kind of debounce.

We could at least add a buf_version check using `lsp.util.buf_versions`, so that it doesn't re-resolve codelenses if the buffer didn't change.

On the other hand, maybe that is premature optimization and not required, we could also leave it as is as a MVP and improve from there if necessary

Reply:
That's a good point. So I guess the only thing we could do is to ensure that only one refresh runs at a time for a given buffer.